py_ecc/bls/api.py

```diff
@@ -80,32 +92,68 @@ def aggregate_pubkeys(pubkeys: Sequence[BLSPubkey]) -> BLSPubkey:
     return G1_to_pubkey(o)
 
 
+def _group_by_messages(
+        message_hashes: Sequence[Hash32],
+        pubkeys: Sequence[BLSPubkey]) -> Iterator[Tuple[Hash32, Tuple[BLSPubkey, ...]]]:
+    if len(pubkeys) != len(message_hashes):
+        raise ValidationError(
+            "len(pubkeys) (%s) should be equal to len(message_hashes) (%s)" % (
+                len(pubkeys), len(message_hashes)
+            )
+        )
+    groups = groupby(itemgetter(0), zip(message_hashes, pubkeys)).items()
+    for message_hash, group in groups:
+        group_pubkeys = tuple(BLSPubkey(pubkey) for _, pubkey in group)
+        yield Hash32(message_hash), group_pubkeys
+
+
 def verify_multiple(pubkeys: Sequence[BLSPubkey],
                     message_hashes: Sequence[Hash32],
                     signature: BLSSignature,
                     domain: int) -> bool:
-    len_msgs = len(message_hashes)
 
-    if len(pubkeys) != len_msgs:
+    o = FQ12.one()
+    for message_hash, group_pubkeys in _group_by_messages(message_hashes, pubkeys):
```

Comment:
@hwwhww This groupby step is actually useless. The use cases in the spec suggest that `bls_verify_multiple` shouldn't group the message_hash

Reply:
Brought back groupby, since it's more like an optimization to save pairings for more general cases.